File: cs445/threading/src/gus/worker/Consumer.java
package gus.worker;

import java.util.*;

public class Consumer implements Runnable {
  private Queue<String> queue;

  public Consumer() {
    this.queue = new LinkedList<String>();
  }

  public synchronized void run() {
    try {
      while (true) {
        String consumable = consume();
        if (consumable == null) {
          System.out.println("I hate waiting :(");
          wait();
        } else {
          System.out.println("I'm a consumer of " + consumable);
        }
      }
    } catch (InterruptedException e) {}
  }

  private synchronized String consume() {
    return this.queue.poll();
  }

  public synchronized void addConsumable(String obj) {
    this.queue.offer(obj);
    notify();
  }
}

File: cs445/threading/ruby/worker/consumer.rb
require 'worker/synchronizable'

#
# ruby worker/producer.rb
class Consumer
  include Synchronizable

  def initialize
    @queue = []
  end

  def consume!
    loop do
      consumable = synchronize { @queue.shift }
      if consumable
        puts "I'm a consumer of #{consumable}"
      else
        puts "I hate waiting :("
        synchronize { wait }
      end
    end
  end

  def add_consumable(consumable)
    synchronize do
      @queue.push(consumable)
      signal
    end
  end
end # Consumer
